### PR TITLE
*: add memory trace

### PR DESCRIPTION
**Pull request by BusyJay**
### What is changed and how it works?

What's Changed:
Instrumenting code to trace memory usage.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- PR to update `pingcap/tidb-ansible`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 
- Instrumenting raftstore to trace memory usage

### CODE REVIEW

File: components/tikv_util/src/memory.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+use std::mem;
+
+pub trait HeapSize {
+    fn heap_size(&self) -> usize;
+}
+
+impl HeapSize for Region {
+    #[inline]
+    fn heap_size(&self) -> usize {
+        self.start_key.len() + self.end_key.len() + mem::size_of::<RegionEpoch>()
```
Comment (hicqu) [inline components/tikv_util/src/memory.rs:16]: Are you sure that `region epoch` is on heap?

Comment (BusyJay) [reply components/tikv_util/src/memory.rs:16]: Yes, it's a protobuf struct. In rust-protobuf, every message field is allocated on heap.

**Comment by BusyJay** [comment]
Most changes are landed in master.

**Pull request closed without merging**